[LPTOCPCI-283] Add documentation for Jira bug creation (#23)
CSPI-QE/ocp-ci-docs: docs/OCP_CI_Tutorials/Reporting/Reporting_Guide.md

[testgrid-config-generator]: https://github.com/openshift/ci-tools/tree/master/cmd/testgrid-config-generator
[kubernetes-test-infra]: https://github.com/kubernetes/test-infra

## Failure Handling (Jira)

### How Failures Are Reported to Jira

Failures are reported to Jira using the [firewatch tool](https://github.com/CSPI-QE/firewatch). This tool is used to react to failures in OpenShift CI jobs. This tool uses a configuration defined for each scenario to help it determine how it should report certain bugs. For a more technical understanding of how to use the tool and build the configuration properly, please see the documentation below:

> - [How to build the configuration](https://github.com/CSPI-QE/firewatch/blob/main/docs/cli_usage_guide.md#defining-the-configuration)
> - [README](https://github.com/CSPI-QE/firewatch/blob/main/README.md)

Reviewer: ```suggestion
- [Getting started](https://github.com/CSPI-QE/firewatch/blob/main/README.md)
- [How to define the configuration](https://github.com/CSPI-QE/firewatch/blob/main/docs/cli_usage_guide.md#defining-the-configuration)
```

> 

Reviewer: ```suggestion
### Example or Example firewatch configuration
```

For the purposes of how this automation works, here is a fairly simple example:

> Each job/scenario in OpenShift CI consists of different steps, for this example we will say our scenario has three steps: `setup`, `test`, and `teardown`. In each of these steps, as far as the automation is concerned, there are two types of faiures: `pod_failure` which means the step's script exited with a non-zero exit code and `test_failure` which means the step generated a JUnit XML file and a failure in the XML was found.

Reviewer: ```suggestion
Each job/scenario in OpenShift CI consists of different steps, for this example we will say our scenario has three steps: `setup`, `test`, and `teardown`. In each of these steps, as far as the automation is concerned, there are two types of failures: `pod_failure` which means the step's script exited with a non-zero exit code and `test_failure` which means the step generated a JUnit XML file and a failure in the XML was found.
```

  {"step": "teardown", "failure_type": "all", "classification": "Lorem Ipsum", "jira_project": "INTEROP"}
]
```

For the sake of this documentation, we will not go very deep into this configuration (again, see the documentation linked above) but this configuration will result in the plain-english rules we outlined earlier. Here is a highly-simplified flowchart of how this works:

```mermaid

flowchart TD
  failure[Test failure in setup step]
  type(Failure type determination)
  rule(Does the step and failure type combo match a rule?)
  matches(Yes)
  no_match(No)
  file_bug(File a bug in the Jira project defined in the rule)
  file_generic_bug(File a bug in the default Jira project)
  comment(Comment on duplicate bug with failure information)

  failure --> type
  type --> 
  rule --> matches & no_match
  matches --> a(Does a duplicate bug exist?) --> d(No) & c(Yes)
  no_match --> b(Does a duplicate bug exist?) --> f(No) & c
  d --> file_bug
  f --> file_generic_bug
  c  --> comment
```

### How To Add Jira Reporting to a Scenario

**If you currently use the ipi-aws workflow:**

1. Ask your PQE contact which Jira project they would like test failures to be reported to

> 2. Modify the scenario to use the firewatch-ipi-aws workflow instead of the ipi-aws workflow

Reviewer: ```suggestion
2. Modify the scenario to use the `firewatch-ipi-aws` workflow instead of the `ipi-aws` workflow
```

3. Add the required environment variables:
   - `FIREWATCH_DEFAULT_JIRA_PROJECT`: This is the Jira project you'd like tickets to be filed to if the failure found does not match any rules. For Interop QE, this will probably be set to `LPTOCPCI`

>    - `FIREWATCH_CONFIG`: Where we define the rules for which tickets get filed where. Please see the []"Defining the Configuration" section of the Firewatch documentation](https://github.com/CSPI-QE/firewatch/blob/main/docs/cli_usage_guide.md#defining-the-configuration) for help defining this variable.

Reviewer: ```suggestion
   - `FIREWATCH_CONFIG`: Where we define the rules for which tickets get filed where. Please see the [How to define the configuration] (https://github.com/CSPI-QE/firewatch/blob/main/docs/cli_usage_guide.md#defining-the-configuration) section of the Firewatch documentation for help defining this variable.
```

   - `FIREWATCH_JIRA_SERVER`: `https://issues.redhat.com`
     - This value always defaults to the stage server to avoid unwanted bugs.

**If you currently use a custom workflow:**

1. Add the `firewatch-report-issues` ref to the end of the post steps in your workflow
2. Ask your PQE contact which Jira project they would like test failures to be reported to
3. Add the required environment variables:
   - `FIREWATCH_DEFAULT_JIRA_PROJECT`: This is the Jira project you'd like tickets to be filed to if the failure found does not match any rules. For Interop QE, this will probably be set to `LPTOCPCI`

>    - `FIREWATCH_CONFIG`: Where we define the rules for which tickets get filed where. Please see the []"Defining the Configuration" section of the Firewatch documentation](https://github.com/CSPI-QE/firewatch/blob/main/docs/cli_usage_guide.md#defining-the-configuration) for help defining this variable.

Reviewer: ```suggestion
   - `FIREWATCH_CONFIG`: Where we define the rules for which tickets get filed where. Please see the [How to define the configuration] (https://github.com/CSPI-QE/firewatch/blob/main/docs/cli_usage_guide.md#defining-the-configuration) section of the Firewatch documentation for help defining this variable.
```
